databricks/koalas/series.py

```diff
@@ -4518,6 +4519,86 @@ def repeat(self, repeats: int) -> "Series":
         else:
             return _col(ks.concat([kdf] * repeats))
 
+    def asof(self, where):
+        """
+        Return the last row(s) without any NaNs before `where`.
+
+        The last row (for each element in `where`, if list) without any
+        NaN is taken.
+
+        If there is no good value, NaN is returned.
+
+        Parameters
+        ----------
+        where : index or array-like of indices
+
+        Returns
+        -------
+        scalar or Series
+
+            The return can be:
+
+            * scalar : when `self` is a Series and `where` is a scalar
+            * Series: when `self` is a Series and `where` is an array-like
+
+            Return scalar or Series
+
+        Notes
+        -----
+        Indices are assumed to be sorted. Raises if this is not the case.
+
+        Examples
+        --------
+        >>> s = ks.Series([1, 2, np.nan, 4], index=[10, 20, 30, 40])
+        >>> s
+        10    1.0
+        20    2.0
+        30    NaN
+        40    4.0
+        Name: 0, dtype: float64
+
+        A scalar `where`.
+
+        >>> s.asof(20)
+        2.0
+
+        For a sequence `where`, a Series is returned. The first value is
+        NaN, because the first element of `where` is before the first
+        index value.
+
+        >>> s.asof([5, 20]).sort_index()
+        5     NaN
+        20    2.0
+        Name: 0, dtype: float64
+
+        Missing values are not considered. The following is ``2.0``, not
+        NaN, even though NaN is at the index location for ``30``.
+
+        >>> s.asof(30)
+        2.0
+        """
+        should_return_series = True
+        if isinstance(self.index, ks.MultiIndex):
+            raise ValueError("asof is not supported for a MultiIndex")
+        if isinstance(where, ks.DataFrame):
+            raise ValueError("where cannot be a DataFrame")
+        if not self.index.is_monotonic_increasing:
+            raise ValueError("asof requires a sorted index")
+        if not is_list_like(where):
+            should_return_series = False
+            where = [where]
+        sdf = self._internal._sdf
+        index_scol = self._internal.index_spark_columns[0]
+        results = [
+            sdf.where(index_scol <= index).select(F.max(self._scol)).head()[0] for index in where
```

Comment:
Wait, this requires to execute the job for every element which is crazy expensive. Can we convert this into single job?

Reply:
I think we did a similar trick before by collecting all maximum values from all columns, convert to pandas. In this way I think you can also pass `where` as series directly.

Reply:
Thanks for the comment.

I just researched about this, Sadly I think we cannot avoid but should run a job for every each items in `where`, because this method requires get the items for every items in `where` from just a single column(`self._scol`) unlike the trick we did before.

In short, in the every step we should get the maximum values less than each items in the single column, and then we a need new single column in the next step.

Reply:
Can you try something like this?

```python
from pyspark.sql import functions as F
spark.range(10).select(
    F.max(F.when(F.col(index_scol) <= index[0], F.col(index_scol)).otherwise(None)),
    F.max(F.when(F.col(index_scol) <= index[1], F.col(index_scol)).otherwise(None))).show()
```
```
+----------------------------------------------+----------------------------------------------+
|max(CASE WHEN (id <= 3) THEN id ELSE NULL END)|max(CASE WHEN (id <= 5) THEN id ELSE NULL END)|
+----------------------------------------------+----------------------------------------------+
|                                             3|                                             5|
+----------------------------------------------+----------------------------------------------+
```

Reply:
> Can you try something like this?
> 
> ```python
> from pyspark.sql import functions as F
> index_scol = "id"
> index = [3, 5]
> spark.range(10).select(
>     F.max(F.when(F.col(index_scol) <= index[0], F.col(index_scol)).otherwise(None)),
>     F.max(F.when(F.col(index_scol) <= index[1], F.col(index_scol)).otherwise(None))).show()
> ```
> 
> ```
> +----------------------------------------------+----------------------------------------------+
> |max(CASE WHEN (id <= 3) THEN id ELSE NULL END)|max(CASE WHEN (id <= 5) THEN id ELSE NULL END)|
> +----------------------------------------------+----------------------------------------------+
> |                                             3|                                             5|
> +----------------------------------------------+----------------------------------------------+
> ```

It works, Thanks!! 👍